### ISSUE TITLE
LIBXSMM: Accelerating Small Matrix Multiplications by Runtime Code Generation

### ISSUE THREAD

**Issue by lingweiyang** [hfp/libxsmm#374]
If the matrix is of single-precision floating point type, what Numbers should Mg and Ng be in the av2 instruction set?

**Comment by hfp** [comment]
We understand you are referring to our publication and you want to know the size of the register blocks in case of AVX2 when using single-precision. In the referred paragraph about AVX2, the register count applies to single-precision (SP) as well, i.e., `Mg` x `Ng` = `{ 16, 12, 8, 4, 2, 1 }` x `{ 1, 2, 3 }`. To summarize, the unit of measure in this sentence is "number of registers" which is the same for single-precision (SP) and double-precision (DP). SP and DP only differ starting with the sentence "As the vector register length in AVX2 is four" (i.e., DP), hence best efficiencies in SP are expected for `M mod 8 == 0`. The latter is because `M` (as opposed to `Mg`) is measured in "number of elements".

**Comment by hfp** [comment]
Feel free to reopen this issue (if your question is not answered), or open another issue if you have more/other questions.

**Comment by lingweiyang** [comment]
what is lengthy auto-tuning phase?

“it describes a hardware-aware code generator which is able to create optimal code without a lengthy auto-tuning phase since just-in-time (JIT) compilation is leveraged. This allows it to exploit problem dimensions which are normally not available.”

**Comment by hfp** [comment]
The sentences says the following:
* The approach targets problem dimensions that were not addressed by, e.g., BLAS, at the time the paper was written.
* The code generator uses JIT to address, i.e., "hard-code", the problem dimensions.
* The code generator does *not* use auto-tuning to determine optimal settings.
* The code generator creates (optimal) code, because it is hardware-aware.

Regarding "hardware-aware", i.e., we have an architectural model of the hardware and we know how to write code optimally, which is leveraged by our code-generator. This is in contrast to approaches that perform blind auto-tuning. In particular, auto-tuning that takes time upfront when building the library.

**Comment by lingweiyang** [comment]
Does the optimal setting refer to  register block? What is the main reason why libxsmm performs better than BLAS, when dealing with small matrices?

**Comment by hfp** [comment]
@lingweiyang would you be willing to share more about your motivation? Perhaps it is an interesting scientific project? I am asking because we were aiminig to publish clear and reproducible information given that the source code is also published and further developed. Have you tried out LIBXSMM already, what is your experience with it?

**Comment by lingweiyang** [comment]
I want to accelerating Small Matrix Multiplications.But I want to do it in assembly language

**Comment by lingweiyang** [comment]
Do you have any good Suggestions?

**Comment by hfp** [comment]
> I want to accelerating Small Matrix Multiplications. But I want to do it in assembly language
If you are interested in LIBXSMM/x86, you can have a look at every kernel you want. There are two ways in LIBXSMM: disassemble a kernel of choice, or generate/print assembly code file using our stand-alone code generator.

> What is the main reason why libxsmm performs better than BLAS, when dealing with small matrices?
Best performance with large GEMMs is related to copy-kernels (effectively hide memory access behind compute/flops), LIBXSMM "reinvented" using direct kernels (no copy). All other reasons are related to x86 properties such as SIB, vector extensions, etc. For batches of small matrix multiplications, it can help to prefetch memory into the cache hierarchy (which LIBXSMM also does). You probably also want to look at cases where the leading dimension(s) is/are larger than the extent of the matrix in question.

You need to study FE and BE of your target CPU and get an idea of what you need to do in order to unlock the full potential. You likely also need to look at a benchmarking methodology for the case(s) you are interested in. For example, which matrix-operands are streamed from (main/DDR-)memoy and which can stay in cache. For instance, excessive unrolling the code when streaming all operands does not make much sense but it looks like it's a good thing if you're using an unrelated benchmark (hot in L1 kernel in particular).

**Comment by lingweiyang** [comment]
Does code size affect the performance of small GEMMs?

**Comment by lingweiyang** [comment]
Does the code size affect the performance of GEMM when it runs only once?

**Comment by hfp** [comment]
Dispatching code, i.e., generating it for the first time or querying it from our code registry adds overhead. In the worst case querying code from our code registry occurs every call of `libxsmm_dgemm` ("auto-dispatch"). Reusing an explicitly dispatched kernel is without any overhead. Querying a kernel from the code registry is the process of using `m`, `n`, `k`, `lda`, `ldb`, `alpha`, and `beta` to determine whether such kernel was previously generated. The code size of the kernel, i.e., the number of Bytes, does not affect performance of code-generation or code-query.

**Comment by lingweiyang** [comment]
For small-scale GEMM performance evaluation, is it run once or multiple times to average?

**Comment by hfp** [comment]
> For small-scale GEMM performance evaluation, is it run once or multiple times to average?

I guess you still refer to our paper? Figure 5 presents the JIT compilation and query times on their own, and figure 6 are execution times without code generation or query time. Figure 7 and 8 are based on reproducer code aiming to model the performance of specific applications (CP2K and SeisSol).

Please note, not only the hardware changed since publication, but also our library. If the background of your question was about current performance, I can recommend to perform new experiments or measurements that match your case. If you want to read a more illustrative article, you can have a look at [https://software.intel.com/en-us/download/parallel-universe-magazine-issue-34-october-2018](https://software.intel.com/en-us/download/parallel-universe-magazine-issue-34-october-2018).

**Comment by lingweiyang** [comment]
Thank you for your answer. I would like to say that for small-scale GEMM, if run multiple times, the code and data will be loaded into the cache, which will affect the evaluation results.

**Comment by hfp** [comment]
True, but this is not specific to SMMs (or our implementation). In fact this is so general, it applies to all code executed on a CPU. I wanted to point out that our kernel/code sizes (number of Bytes) are far from hitting I$ size. Our kernels are not heavily unrolled (they could be unrolled more). However, unrolling heavily only affects when running data hot in L1$ which is barely a realistic case (it can be but unlikely). It is very important to understand what is measured by a specific benchmark like which operands are streamed from memory and which can stay in (data-)cache (some of this is determined by `beta`, e.g., beta=1 means to accumulate into the same matrix which in turn can likely stay in L1$). It is further important to look at leading dimensions as LIBXSMM kernels can handle the latter in a way unlikely available to other codes.

**Comment by lingweiyang** [comment]
I don't think SMM evaluation should be run multiple times. What do you think?
